### PR TITLE
[BEAM-14117] Unvendor bytebuddy dependency

### PR DESCRIPTION
**Pull request by cushon**
This change unvendors bytebuddy, as discussed in [BEAM-14117](https://issues.apache.org/jira/browse/BEAM-14117)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by kennknowles** [comment]
run java postcommit

**Comment by kennknowles** [comment]
run flink validatesrunner

**Comment by kennknowles** [comment]
run spark validatesrunner

**Comment by kennknowles** [comment]
run samza validatesrunner

**Comment by kennknowles** [comment]
run dataflow validatesrunner

**Comment by cushon** [comment]
R: @kennknowles

**Comment by kennknowles** [comment]
Thread at https://lists.apache.org/thread/sdp7tpvy5n1jlk3g8wqqp40dlyzp5llv did not encounter any major objections.

**Comment by kennknowles** [comment]
The tests are super flaky, but I think Jenkins runs against the HEAD commit of the PR so it will need a rebase to get any recent improvements in test flakiness. Let me investigate a little bit and I'll comment when I have news.

File: sdks/java/core/build.gradle [hunk 1]
```diff
@@ -28,12 +28,17 @@ applyJavaNature(
     dependencies {
       include(dependency("org.apache.commons:.*"))
       include(dependency(library.java.antlr_runtime))
+      include(dependency(library.java.byte_buddy))
```
Comment (kennknowles) [inline sdks/java/core/build.gradle:31]: I don't think we should bundle bytebuddy in here, right? It should be a residual dependency that is resolved according to everyone's poms?

Comment (kennknowles) [reply sdks/java/core/build.gradle:31]: In other words, it should become "just like any other dependency"

Comment (cushon) [reply sdks/java/core/build.gradle]: Done, thanks

**Review by kennknowles — APPROVED** [review]
Perfect!

**Comment by kennknowles** [comment]
Now I just want to poke the tests until they are green, rather than ignoring red signals, even if they seem irrelevant.

**Comment by cushon** [comment]
> Any idea why the regular non-shadow dependency isn't getting resolved correctly there?

The tests are making more progress after adding an explicit dep to `sdks/java/harness/jmh/build.gradle`.

**Comment by lukecwik** [comment]
> > Any idea why the regular non-shadow dependency isn't getting resolved correctly there?
> 
> The tests are making more progress after adding an explicit dep to `sdks/java/harness/jmh/build.gradle`.

Adding an explicit dep doesn't seem to be the way to go since it is technically a transitive dependency.

**Comment by cushon** [comment]
> Adding an explicit dep doesn't seem to be the way to go since it is technically a transitive dependency.

Thanks, any suggestions about what the right way to go would be?

File: sdks/java/harness/jmh/build.gradle [hunk 2]
```diff
@@ -34,6 +34,7 @@ dependencies {
     implementation project(path: ":sdks:java:harness", configuration: "shadow")
     implementation project(":runners:java-fn-execution")
     runtimeOnly library.java.slf4j_jdk14
+    runtimeOnly library.java.byte_buddy
```
Comment (kennknowles) [inline sdks/java/harness/jmh/build.gradle:37]: I'm not sure how the failure occurred, but based on grep and IWYU this should just not be here. I understand you added it in response to an issue. It sounds like there must be an error in some other module's IWYU compliance, and presumably is suppressed? (just a guess; I don't know)

**Comment by cushon** [comment]
I did some more debugging, and my understanding is that this configuration is effectively removing the `implementation` dep:

https://github.com/apache/beam/blob/f2e41be4880ef47d86d975c38f353896f70365bd/sdks/java/core/build.gradle#L27-L35

because when `shadowClosure` is specified `BeamModulePlugin` removes all of the implementation deps:

https://github.com/apache/beam/blob/f2e41be4880ef47d86d975c38f353896f70365bd/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy#L808-L810

Hacking up `BeamModulePlugin` to add a hard-coded implementation dep on bytebuddy fixes the build errors I was seeing, which seems to be consistent with that theory:

```diff
diff --git a/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy b/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy
index 791277a7e1..30b94333e0 100644
--- a/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy
+++ b/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy
@@ -496,6 +496,7 @@ class BeamModulePlugin implements Plugin<Project> {
     def testcontainers_version = "1.16.3"
     def arrow_version = "5.0.0"
     def jmh_version = "1.34"
+    def byte_buddy_version = "1.12.9"

     // A map of maps containing common libraries used per language. To use:
     // dependencies {
@@ -541,7 +542,7 @@ class BeamModulePlugin implements Plugin<Project> {
         aws_java_sdk2_utils                         : "software.amazon.awssdk:utils:$aws_java_sdk2_version",
         bigdataoss_gcsio                            : "com.google.cloud.bigdataoss:gcsio:$google_cloud_bigdataoss_version",
         bigdataoss_util                             : "com.google.cloud.bigdataoss:util:$google_cloud_bigdataoss_version",
-        byte_buddy                                  : "net.bytebuddy:byte-buddy:1.12.9",
+        byte_buddy                                  : "net.bytebuddy:byte-buddy:$byte_buddy_version",
         cassandra_driver_core                       : "com.datastax.cassandra:cassandra-driver-core:$cassandra_driver_version",
         cassandra_driver_mapping                    : "com.datastax.cassandra:cassandra-driver-mapping:$cassandra_driver_version",
         cdap_api                                    : "io.cdap.cdap:cdap-api:$cdap_version",
@@ -1063,7 +1064,15 @@ class BeamModulePlugin implements Plugin<Project> {
         // This contains many improved annotations beyond javax.annotations for enhanced static checking
         // of the codebase. It is runtime so users can also take advantage of them. The annotations themselves
         // are MIT licensed (checkerframework is GPL and cannot be distributed)
-        implementation "org.checkerframework:checker-qual:$checkerframework_version"
+
+        def implementationDeps = [
+          "org.checkerframework:checker-qual:$checkerframework_version",
+          "net.bytebuddy:byte-buddy:$byte_buddy_version",
+        ]
+
+        implementationDeps.each { dep ->
+          implementation dep
+        }
       }

       // Defines Targets for sonarqube analysis reporting.
```

What do you recommend? It seems like there are no existing examples of implementation deps that aren't included in the shadow closure. Does it make sense to include it in the shadow closure after all? Is there a more principled change to `BeamModulePlugin` to allow it to be configured to preserve some implementation deps?

**Comment by lukecwik** [comment]
> I did some more debugging, and my understanding is that this configuration is effectively removing the `implementation` dep:
> 
> https://github.com/apache/beam/blob/f2e41be4880ef47d86d975c38f353896f70365bd/sdks/java/core/build.gradle#L27-L35
> 
> because when `shadowClosure` is specified `BeamModulePlugin` removes all of the implementation deps:
> 
> https://github.com/apache/beam/blob/f2e41be4880ef47d86d975c38f353896f70365bd/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy#L808-L810
> 
> Hacking up `BeamModulePlugin` to add a hard-coded implementation dep on bytebuddy fixes the build errors I was seeing, which seems to be consistent with that theory:
> 
> ```diff
> diff --git a/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy b/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy
> index 791277a7e1..30b94333e0 100644
> --- a/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy
> +++ b/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy
> @@ -496,6 +496,7 @@ class BeamModulePlugin implements Plugin<Project> {
>      def testcontainers_version = "1.16.3"
>      def arrow_version = "5.0.0"
>      def jmh_version = "1.34"
> +    def byte_buddy_version = "1.12.9"
> 
>      // A map of maps containing common libraries used per language. To use:
>      // dependencies {
> @@ -541,7 +542,7 @@ class BeamModulePlugin implements Plugin<Project> {
>          aws_java_sdk2_utils                         : "software.amazon.awssdk:utils:$aws_java_sdk2_version",
>          bigdataoss_gcsio                            : "com.google.cloud.bigdataoss:gcsio:$google_cloud_bigdataoss_version",
>          bigdataoss_util                             : "com.google.cloud.bigdataoss:util:$google_cloud_bigdataoss_version",
> -        byte_buddy                                  : "net.bytebuddy:byte-buddy:1.12.9",
> +        byte_buddy                                  : "net.bytebuddy:byte-buddy:$byte_buddy_version",
>          cassandra_driver_core                       : "com.datastax.cassandra:cassandra-driver-core:$cassandra_driver_version",
>          cassandra_driver_mapping                    : "com.datastax.cassandra:cassandra-driver-mapping:$cassandra_driver_version",
>          cdap_api                                    : "io.cdap.cdap:cdap-api:$cdap_version",
> @@ -1063,7 +1064,15 @@ class BeamModulePlugin implements Plugin<Project> {
>          // This contains many improved annotations beyond javax.annotations for enhanced static checking
>          // of the codebase. It is runtime so users can also take advantage of them. The annotations themselves
>          // are MIT licensed (checkerframework is GPL and cannot be distributed)
> -        implementation "org.checkerframework:checker-qual:$checkerframework_version"
> +
> +        def implementationDeps = [
> +          "org.checkerframework:checker-qual:$checkerframework_version",
> +          "net.bytebuddy:byte-buddy:$byte_buddy_version",
> +        ]
> +
> +        implementationDeps.each { dep ->
> +          implementation dep
> +        }
>        }
> 
>        // Defines Targets for sonarqube analysis reporting.
> ```
> 
> What do you recommend? It seems like there are no existing examples of implementation deps that aren't included in the shadow closure. Does it make sense to include it in the shadow closure after all? Is there a more principled change to `BeamModulePlugin` to allow it to be configured to preserve some implementation deps?

You should declare byte_buddy within the shadow configuration as per https://github.com/apache/beam/blob/f2e41be4880ef47d86d975c38f353896f70365bd/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy#L811

This will make it a compile time dependency for the source set and it will be exposed in the pom.xml file and should become a transitive dependency.

**Comment by cushon** [comment]
> You should declare byte_buddy within the shadow configuration

Thanks so much! Done.

**Comment by cushon** [comment]
I think all of the open issues have been resolved:

* the byte_buddy dep is now in the shadow configuration
* the CI is green (I had to retry a few times but this didn't require any additional fixes, I think some of the tests might be flaky)

Let me know if there's any additional feedback!

**Review by lukecwik — APPROVED** [review]

**Comment by lukecwik** [comment]
Fixes https://github.com/apache/beam/issues/21519

**Pull request merged**